library/common/engine.cc

```diff
@@ -94,14 +95,10 @@ Engine::~Engine() {
 }
 
 void Engine::recordCounter(std::string elements, uint64_t count) {
-  if (server_) {
+  if (server_ && client_scope_) {
     server_->dispatcher().post([this, elements, count]() -> void {
```

Comment:
why the post() here rather than just incrementing the counter directly at the call-site?

counters are thread-safe, at least once they are looked up.

Maybe the issue is that the caller is in some thread that Envoy does not know about, so the TLS lookups underneath fail? In that case can you comment?

Reply:
The issue is exactly what you suggest - this call is coming from a random application thread, Envoy has no knowledge of it, and there's no TLS state associated with it. I can definitely add a comment, though that's actually a common theme at this layer of the API.

Reply:
No worries just add a comment explaining the non-obvious delay.

Reply:
On the other hand, if you see a way we could accomplish this without the dispatch in a thread-safe manner, would certainly be interested in hearing more.

Reply:
I don't think you can without changes to ThreadLocalStore...so we'll have to be content with a comment.

Also when you establish an API for creation of a Counter object for callers, the same issue will be there. You will have to populate the underlying Counter objects from a dispatch.

I'll defer to @mattklein123 on what the best TLS strategy should be for accessing stat by name outside the context of worker threads. But in the current code I think it has to be done via post.